be/etp-backend/src/main/java/com/psa/TradingEngine.java

```diff
@@ -13,145 +13,199 @@
-    } catch (SQLException e) {
-      throw new RuntimeException(e);
+
+    private void updateLastPrice(String symbol, double price) throws SQLException {
+        // Update the price if the symbol exists, if not - insert a new record
+        String sql = "INSERT INTO public.last_price (symbol, price) " +
+                "VALUES (?, ?) " +
+                "ON CONFLICT (symbol) DO UPDATE " +
+                "SET price = EXCLUDED.price";
+
+        try (PreparedStatement pstmt = db.prepareStatement(sql)) {
+            pstmt.setString(1, symbol);
+            pstmt.setDouble(2, price);
+
+            pstmt.executeUpdate();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
     }
-  }
-
-  private void updateOrder(OrderEntity order, int filled) throws SQLException {
-    String sql = "UPDATE public.order SET quantity_filled = quantity_filled + ?, status = ? WHERE id = ?";
-    try (PreparedStatement pstmt = db.prepareStatement(sql)) {
-      pstmt.setInt(1, filled);
-      pstmt.setString(2, (
-            order.quantity_filled() +
-            order.quantity_cancelled() +
-            filled
-            ) >= order.quantity() ? "COMPLETE" : order.status());
-      pstmt.setString(3, order.id());
-      pstmt.executeUpdate();
-    } catch (SQLException e) {
-      throw new RuntimeException(e);
+
+    // Create a symbol record, possibly put in scanOrderBook, where would the status come from?
+    private void createSymbolRecord(String symbol, String status) throws SQLException {
+        String sql = "INSERT INTO public.symbol (symbol, status) " +
+                "VALUES (?, ?) " +
+                "ON CONFLICT (symbol) DO NOTHING";
+        try (PreparedStatement pstmt = db.prepareStatement(sql)) {
+            pstmt.setString(1, symbol);
+            pstmt.setString(2, status);
+
+            pstmt.executeUpdate();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    // Have put the endpoint in ExchangeGrpcImpl, leaving this here for potential later use.
+    private void getSymbols() throws SQLException {
+        String sql = "SELECT * FROM public.symbol";
+        try (PreparedStatement pstmt = db.prepareStatement(sql)) {
+            pstmt.executeUpdate();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
     }
-  }
-
-  private void updateLastPrice(String symbol, double price) throws SQLException {
-    // Update the price if the symbol exists, if not - insert a new record
-    String sql = "INSERT INTO public.last_price (symbol, price) " +
-            "VALUES (?, ?) " +
-            "ON CONFLICT (symbol) DO UPDATE " +
-            "SET price = EXCLUDED.price";
-
-    try (PreparedStatement pstmt = db.prepareStatement(sql)) {
-      pstmt.setString(1, symbol);
-      pstmt.setDouble(2, price);
-
-      pstmt.executeUpdate();
-    } catch (SQLException e) {
-      throw new RuntimeException(e);
+
```

Comment:
Oh nice. You didn't have to go this far. But this is precisely correct. We will need this function eventually when we write the stock status manager mechanism. And exactly, we would not expose this function through any public endpoint. You mention an 'admin' user—that's one of the correct ideas. The way this is actually done is, see how in exchange.proto there's `service Exchange { ... }`? Well we would add a second one, `service ExchangeInternal { ... }` that has endpoints only meant for internal use. That's _one_ correct way. The other correct way is, if the symbol status manager is just some code in this program, then there's no reason for any endpoints to exist. For example, if in Java we just set Timeouts so that at the next Monday-Friday 9:30am we change all symbol statuses to OPEN, and on the next Monday-Friday 4pm we change all symbol statuses to CLOSED, then this logic can just call updateSymbolStatus directly, without need for an endpoint. But, of the 2 approaches, your choice (the first one) is more correct. That's because there are actually many statuses and all kinds of rules. Some are by time. Some are by situation, _e.g._ an admin panel that allows admins to HALT a specific stock at any time. So it's best to have a separate service that manages symbol statuses through an ExchangeInternal client. But don't go any further here for now, you've done enough.